actor/actor.go

```diff
@@ -39,64 +36,176 @@ type Server interface {
 	// Type is defined by user
 	Type() string
 	// SetStateManager is impl by ServerImplBase to inject StateManager to this actor instance
+	// Deprecated: SetStateManager is deprecated in favour of SetStateManagerContext.
 	SetStateManager(StateManager)
 	// SaveState is impl by ServerImplBase, It saves the state cache of this actor instance to state store component by calling api of daprd.
 	// Save state is called at two places: 1. On invocation of this actor instance. 2. When new actor starts.
 	SaveState() error
+
+	WithContext() ServerContext
+}
+
+// ServerContext is the interface that would be impl by user's actor server with ServerImplBaseCtx
+/*
+Actor user should only impls func Type() string, and his user-defined-method, Other function could be impl by
+combining  ServerImplBaseCtx.
+*/
+type ServerContext interface {
+	// ID is impl by ServerImplBase. It can be called by user defined actor function to get the actor ID of it's instance.
+	ID() string
+	// SetID is impl by ServerImplBase. It is called by actor container to inject actor ID of the instance, and should
+	// not called by user
+	SetID(string)
+	// Type is defined by user
+	Type() string
+	// SetStateManager is impl by ServerImplBase to inject StateManager to this actor instance
+	SetStateManager(StateManagerContext)
+	// SaveState is impl by ServerImplBase, It saves the state cache of this actor instance to state store component by calling api of daprd.
+	// Save state is called at two places: 1. On invocation of this actor instance. 2. When new actor starts.
+	SaveState(context.Context) error
 }
 
 type ReminderCallee interface {
 	ReminderCall(string, []byte, string, string)
 }
 
-type Factory func() Server
+type (
+	Factory        func() Server
+	FactoryContext func() ServerContext
+)
 
+// Deprecated: ServerImplBase is deprecated in favour of ServerImplBaseCtx.
 type ServerImplBase struct {
 	stateManager StateManager
+	ctx          ServerImplBaseCtx
+}
+
+type ServerImplBaseCtx struct {
+	stateManager StateManagerContext
 	once         sync.Once
 	id           string
+	lock         sync.RWMutex
 }
 
+// Deprecated: Use ServerImplBaseCtx instead.
 func (b *ServerImplBase) SetStateManager(stateManager StateManager) {
+	b.ctx.lock.Lock()
+	defer b.ctx.lock.Unlock()
 	b.stateManager = stateManager
+	b.ctx.stateManager = stateManager.WithContext()
```

Comment:
Does this create a race condition where two contexts can be set while the invocation did not take place yet? Is there a lock somewhere surrounding WithContext() and SaveState()?

The lock here will avoid another race condition instead.

Reply:
Thanks, there can be. I'll add another lock to `ServerImpBase`.